reaper.go

```diff
@@ -38,22 +38,39 @@ type ReaperProvider interface {
 // NewReaper creates a Reaper with a sessionID to identify containers and a provider to use
 // Deprecated: it's not possible to create a reaper anymore.
 func NewReaper(ctx context.Context, sessionID string, provider ReaperProvider, reaperImageName string) (*Reaper, error) {
-	return newReaper(ctx, sessionID, provider, WithImageName(reaperImageName))
+	return reuseOrCreateReaper(ctx, sessionID, provider, WithImageName(reaperImageName))
 }
 
-// newReaper creates a Reaper with a sessionID to identify containers and a provider to use
-func newReaper(ctx context.Context, sessionID string, provider ReaperProvider, opts ...ContainerOption) (*Reaper, error) {
+// reuseOrCreateReaper returns an existing Reaper instance if it exists and is running. Otherwise, a new Reaper instance
+// will be created with a sessionID to identify containers and a provider to use
+func reuseOrCreateReaper(ctx context.Context, sessionID string, provider ReaperProvider, opts ...ContainerOption) (*Reaper, error) {
 	mutex.Lock()
 	defer mutex.Unlock()
-	// If reaper already exists re-use it
-	if reaper != nil {
-		return reaper, nil
+	// If reaper already exists and healthy, re-use it
+	if reaperInstance != nil {
+		// Verify this instance is still running by checking state.
+		// Can't use Container.IsRunning because the bool is not updated when Reaper is terminated
```

Comment:
Meanwhile the CI runs (tests will pass) can you elaborate a bit on this comment? Do you think that bool should be updated in consequence (probably here? in a follow-up?

Reply:
So the `Container.IsRunning` is simply a boolean flag (https://github.com/testcontainers/testcontainers-go/blob/fb646713ffeaede8472fb1dd28dce495c6f3ab6f/docker.go#L63) that gets updated when a container is stopped / terminated (https://github.com/testcontainers/testcontainers-go/blob/fb646713ffeaede8472fb1dd28dce495c6f3ab6f/docker.go#L233). And while the Reaper is a 'regular' container, it is never terminated via the `Stop()` or `Terminate()` methods. Instead, it terminates automatically by itself by closing the connection to it (https://github.com/testcontainers/testcontainers-go/blob/fb646713ffeaede8472fb1dd28dce495c6f3ab6f/docker.go#L241).
Which makes the `IsRunning()` unreliable for the Reaper container as it would never return `false`.

That's why I chose to poll Docker and get the state from there, as it should be reliable.

What I think would be better, is if the `IsRunning()` method (https://github.com/testcontainers/testcontainers-go/blob/fb646713ffeaede8472fb1dd28dce495c6f3ab6f/docker.go#L89) gets updated to what I'm doing below; checking from Docker if the state is running or not. That would make it reliable and not depend on a boolean flag internally updated.
I did not want to touch this code in the PR because I don't know the history behind it. Perhaps `IsRunning` is regularly checked and would be relatively heavy on the Docker calls?

Reply:
I see, thanks for such a clear explanation, now I see what you mean.

I think checking the state from Docker would be more accurate than the one we have now, and I understand your concerns about performance if that method is called many times. I've checked their usages and are just a few of them at this moment (particularly worried on the GenericContainer function calling it).

Adding that code, we could also take the situation to a place where multiple goroutines in tests are trying to read the container state at the same time, and one decides to see it as not running, and the next one as running, which could be worse.

I'd keep this simple as in your submission, and evaluate if we need to check for the state more frequently.